make sure all files are readable before packaging for Lambda

skurfer opened the pull request:
As described in the commit message, this will create a temporary copy of package files so permissions on the originals aren’t modified. I don’t *think* we need to do that dance for the library source (since you generally only read from them to begin with), but let me know if you think we should.

kapilt commented:
Rather than making a copy of the files, it would be preferrable to create a zip info and set the attrs when adding to the archive.

skurfer commented:
I looked for a way to do that (and at the `ZipInfo` documentation specifically), but it wasn’t obvious. I see some examples out there for it, though. I’ll update this.

kapilt commented:
sounds good, there's a few good extant links, we do it as well in mu.py for files we generate re permission bit set

skurfer commented:
Updated and rebased.

There’s an issue with coverage on the code to handle directories. I’ll look into that, but don’t let it hold up your review. We can address that as part of the overall coverage improvements.

kapilt commented on c7n/mu.py:
```diff
@@ -105,9 +106,23 @@ def create(self):
-                        os.path.join(root, f),
-                        os.path.join(arc_prefix, f))
+                    f_path = os.path.join(root, f)
+                    info = zipfile.ZipInfo(os.path.join(arc_prefix, f))
```
rather than copying and pasting the lines, preference would be to extract out to a utility function that's invoked from both sites.

skurfer commented:
New commit. This is ready for review.

kapilt commented on c7n/mu.py:
```diff
@@ -119,18 +120,27 @@ def create(self):
+        info.date_time = datetime.timetuple(
+            datetime.fromtimestamp(timestamp)
+        )
+        fp = open(src, 'rb')
```
any not to use context manager on open?

skurfer replied:
Seemed like overkill for one line, but it has the added benefit of cleaning up if anything goes wrong, so I’ll change it.

kapilt commented on c7n/mu.py:
```diff
@@ -119,18 +120,27 @@ def create(self):
             dest = zinfo(dest)
         # see zinfo function for some caveats
         assert not self._closed, "Archive closed"
+        dest.external_attr = 0444 << 16L
```
this makes the line in the zinfo method redundant as well. it would be good to have a test verifying this is behaving as intended.

kapilt commented on c7n/mu.py:
```diff
@@ -119,18 +120,27 @@ def create(self):
-        self._zip_file.write(src, dest)
+        info = zipfile.ZipInfo(dest)
+        timestamp = os.path.getmtime(src)
+        info.date_time = datetime.timetuple(
```
time.localtime is less churn on temp objects

its worth reviewing the std lib implementation
https://github.com/python/cpython/blob/master/Lib/zipfile.py#L473

skurfer replied:
I thought it would be nice to preserve the date, but since the archive will only ever be used on Lamba, it’s probably not worth the overhead. I’ll take that out.

skurfer commented:
Added some commits and rebased.

kapilt commented:
lgtm
